### PR TITLE
config: allow globs in federation_domain_whitelist

### LINKED ISSUE #7346: federation_domain_whitelist wildcard domain support …
It would be good if `federation_domain_whitelist` would support wildcard domains, as for example `federation_certificate_verification_whitelist` does support it.

```
#federation_certificate_verification_whitelist:
#  - lon.example.com
#  - *.domain.com
#  - *.onion

```

### PR DESCRIPTION
**Pull request by 3nprob**
Resolves #7346

Allows to use glob patterns in `federation_domain_whitelist`.

```
federation_domain_whitelist:
  - *.example.com
  # [...]
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by 3nprob** [comment]
```
Signed-off-by: 3nprob <git@3n.anonaddy.com>
```

**Comment by 3nprob** [comment]
Using `dict` as underlying makes less sense from a performance perspective after this change - I can change it to something more appropriate, shout out if you have any favorite datastructures for the job (and no we're not doing bloom filters :stuck_out_tongue_winking_eye: )

**Comment by 3nprob** [comment]
Broke out `DomainGlobSet` (naming improvement suggestions welcome) to share implementation.

File: synapse/config/_util.py [hunk 1]
```diff
@@ -61,3 +63,20 @@ def json_error_to_config_error(
...
+class DomainGlobSet:
+    def __init__(self, globs: list):
+        self._globs = []
+        for entry in globs:
+            try:
+                self._globs.append(glob_to_regex(entry.encode("ascii").decode("ascii")))
+            except UnicodeEncodeError:
+                raise ConfigError("IDNA domain names are not allowed: %s " % (entry,))
+
+    @lru_cache(maxsize=100000)
```
Comment (richvdh) [inline synapse/config/_util.py:77]: suggest adding a comment here recording your notes about how `@lrucache` is good enough for now, but a future optimisation could be to use a tree.

**Review by richvdh — CHANGES_REQUESTED** [review]
This looks like a sensible approach, thank you!

Some requests:
 * please could you give `DomainGlobSet` a docstring. What does it do, how should it be used?
 * please could you make sure that the new functionality gets some test coverage, to make sure it stays working. We'll need to test both inbound and outbound traffic: I suggest you look at `FederationClientTests` and `FederationServerTests` respectively.
 * please make sure the change to the config setting is documented in `docs/usage/configuration/config_documentation.md`.

File: synapse/crypto/context_factory.py [hunk 2]
```diff
@@ -142,14 +142,9 @@ def get_options(self, host: bytes) -> IOpenSSLClientConnectionCreator:
...
-
-        # Check if we've disabled certificate verification for this host
-        if self._should_verify:
-            for regex in self._federation_certificate_verification_whitelist:
-                if regex.match(ascii_host):
-                    should_verify = False
-                    break
+        should_verify = self._should_verify and not (
+            ascii_host in self._federation_certificate_verification_whitelist
+        )
```
Comment (richvdh) [inline synapse/crypto/context_factory.py:147]: sonarcloud wants you to write `x not in y` rather than `not (x in y)`

**Comment by clokep** [comment]
I'm going to close this as the author has not responded; @3nprob if you're interested in continuing to work on this please let us know and respond to any outstanding feedback. Thank you!

**Comment by 3nprob** [comment]
Apologies for dropping the ball here.

FWIW the closed version still operates fine on a fork - anyone is free to extend this and resubmit it as a new PR. I may or may not get around to that some day if nobody else does.

**Pull request closed without merging**